### PR TITLE
Add numeric_limits for MLFloat16 and BFloat16

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description
* Add std::numeric_limits for MLFloat16 and BFloat16.
* Update some comments in csharp ORTFloat16.shared.cs.
* Add unit tests (including Clip)

Note that the canonical NaN is not consistent in C++ and C#. C# uses negative quiet NaN as canonical NaN, while C++ uses positive quiet NaN. The choice of CSharp Float16.NaN is to be consistent with System.Half.NaN.

FP16 data returns from CUDA might have 7FFF as NaN; FP16 data from CPU provider might have 0x7E00 as NaN. Anyway there is no consistent canonical NaN in ORT right now. Because all these NaNs are aligned with IEEE spec, there shall not an issue in downstream.

### Motivation and Context
std::numeric_limits is used in codebase but not defined for MLFloat16 and BFloat16. It causes some bugs like https://github.com/microsoft/onnxruntime/issues/21957 introduced by https://github.com/microsoft/onnxruntime/pull/21493.

### CODE REVIEW

**Review by snnn — APPROVED** [review]

File: csharp/src/Microsoft.ML.OnnxRuntime/OrtFloat16.shared.cs [hunk 1]
```diff
@@ -248,17 +249,17 @@ internal static float CreateSingle(bool sign, byte exponent, uint significand)
...
         /// Float16 Negative Infinity value
         /// </summary>
-        public static Float16 NegativeInfinity => new Float16(NegativeInfinityBits);      // -1.0 / 0.0
+        public static Float16 NegativeInfinity => new Float16(NegativeInfinityBits);
 
         /// <summary>
         /// Float16 NaN
         /// </summary>
-        public static Float16 NaN => new Float16(NegativeQNaNBits);                       //  0.0 / 0.0
+        public static Float16 NaN => new Float16(PositiveQNaNBits);                       //  quiet NaN
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/OrtFloat16.shared.cs:262]: The issue with this change, is that now it is going to diverge with the corresponding `System.Half` class introduced in the later version of .NET framework.
It is a good thing to be consistent in the product implementation across the interfaces, but C# coders may disagree.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Half.cs#L95

Said that, IsNaN() would still work properly.

The alternative proposal to fix C++ code to test for Nans in a more comprehesive manner.

Comment (snnn) [reply csharp/src/Microsoft.ML.OnnxRuntime/OrtFloat16.shared.cs:262]: C++ on Linux also has a builtin float16 type: https://gcc.gnu.org/onlinedocs/gcc/Half-Precision.html

**Comment by yuslepukhin** [comment]
Do we want to add a test for Clip or this is separate?

File: csharp/src/Microsoft.ML.OnnxRuntime/OrtFloat16.shared.cs (hunk 1 above)
Comment (yuslepukhin) [reply csharp/src/Microsoft.ML.OnnxRuntime/OrtFloat16.shared.cs]: I thinking about this scenario.

1) when results returned with NaN from C++ C# would be able to test it with IsNaN() no issues.

2) if the user feeds data from C# to C++ and it has NaN defined as NegativeNaN, would C++ be able to detect that properly. If so, no issues.

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Pull request merged**